### ISSUE TITLE
Test build with ipfs add ignore support

### ISSUE THREAD

**Issue by bmann** [fission-suite/go-ipfs#2]
@corntoole made a build which can be downloaded here https://send.firefox.com/download/f624ef135fb70d3a/#Y8s_VYhnovph3sL7Cn-Hag

See https://github.com/corntoole/go-ipfs-stack for where these things are being staged

> this will require 3 PRs. go-ipfs-files 1st, then go-ipfs-cmds, and lastly, go-ipfs. I don't think we can revive the old PRs.
> the feature branch for [go-ipfs#4103](https://github.com/ipfs/go-ipfs/pull/4013) seems to have been deleted; go-ipfs-cmdkit has been rolled into go-ipfs-cmds; whereas, the SerialFile type has been moved into go-ipfs-files

**Comment by icidasset** [comment]
The `.gitignore` functionality works great 👌 

I'm having a few issues with the `--ignore` flag though:
```
⚡ ipfs-ignore add . -r -i *.png
added QmUaRffjF2jZpYJ4hZBbxXuwYimDuQPhBeJxjSoDzMZNFT Drive Screenshots/drive-index-large-dark.png
added QmTJJCkWi52uerxX9UFemX5KKPhE1aTPzwt8dsUBERzB18 Drive Screenshots/drive-index-large-light.png
added QmUXfouoNYyqWVfbwmTibSaEzbMmNRdBP27UVYJ4TayBDS Drive Screenshots/drive-index-small-dark.png
added Qme5fRH1zR133FN91SkkbGKboLnSE4WnWKEPxmxUdtzjqH Drive Screenshots/drive-index-small-light.png
added QmV9zt1UEtZUqFymDfgW417iEms1KzwpD7gH843xrwane6 Drive Screenshots/drive-preview-large-dark.png
added QmdKQbdVgBRsse2hoeq5mL9maSQJp8vzAGi9XyZbyMktmG Drive Screenshots/drive-preview-large-light.jpg
added QmTKSb1TuCBRtH5DmYGG5tdoX2xRfuYu7NbkQ4Gf1rexNq Drive Screenshots/drive-preview-small-dark.png
added Qmb3kRU7RAhE8Ei1FBNr6rUU9oVKqgEtHtDz53i9rgSmmn Drive Screenshots/drive-preview-small-light.jpg
added QmSBXWr6UQUw5ZmHaLozzRphAy7TjdYNgBgq7fynMUGeNM Drive Screenshots
added QmUaRffjF2jZpYJ4hZBbxXuwYimDuQPhBeJxjSoDzMZNFT drive-index-large-dark.png
added QmTJJCkWi52uerxX9UFemX5KKPhE1aTPzwt8dsUBERzB18 drive-index-large-light.png
added QmUXfouoNYyqWVfbwmTibSaEzbMmNRdBP27UVYJ4TayBDS drive-index-small-dark.png
added Qme5fRH1zR133FN91SkkbGKboLnSE4WnWKEPxmxUdtzjqH drive-index-small-light.png
added QmV9zt1UEtZUqFymDfgW417iEms1KzwpD7gH843xrwane6 drive-preview-large-dark.png
added QmTKSb1TuCBRtH5DmYGG5tdoX2xRfuYu7NbkQ4Gf1rexNq drive-preview-small-dark.png
```

1. The lines that are __not__ prefixed by `Drive Screenshots` shouldn't be there (ie. the lines after `QmSBXWr6UQUw5ZmHaLozzRphAy7TjdYNgBgq7fynMUGeNM`). I compared this output to the output of the regular ipfs-cli.
2. The prefixed lines are not filtered (there's still jpg files in the list)
2. This ignore flag behaves like a `--keep` flag, so either we should rename this flag or fix the behaviour.

**Comment by corntoole** [comment]
Thanks for the feedback. I'm not seeing that behavior. Did you build or download the binary you tested with, @icidasset ?

Initially, I thought that the `-i` / `--ignore` rules were being applied as simple strings, but they are applied as regular expressions. So file globs should work.

Here are some paste dumps from my ad-hoc testing on my box.

## Test 1 - should not add any PNG files
```
corntoole at cornetopia1 in /tmp/ipfs-add-test
$ ipfs add -r -i *.png ./another_dir
added QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn another_dir
 0 B / ? [---------------------------------------------------------------------------=]   0.00%
```

## Test 2 - should all files under another_dir
```

corntoole at cornetopia1 in /tmp/ipfs-add-test
$ ipfs add -r ./another_dir
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH another_dir/a.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH another_dir/b.png
added QmQaaNSYM2RP26aoSVhgyJWEc7kMpSqaD2uKRqsZjLsK95 another_dir
 0 B / ? [---------------------------------------------------------------------------=]   0.00%
```


## Test 3 - should not add any PNG files matching the pattern `foo*.png`
```
corntoole at cornetopia1 in /tmp/ipfs-add-test
$ mkdir another_other_dir

corntoole at cornetopia1 in /tmp/ipfs-add-test/another_other_dir
$ touch c.png

corntoole at cornetopia1 in /tmp/ipfs-add-test/another_other_dir
$ touch d.png

corntoole at cornetopia1 in /tmp/ipfs-add-test/another_other_dir
$ touch foo_a.png

corntoole at cornetopia1 in /tmp/ipfs-add-test/another_other_dir
$ touch foo_b.png

corntoole at cornetopia1 in /tmp/ipfs-add-test/another_other_dir
$ touch foo_c.png

corntoole at cornetopia1 in /tmp/ipfs-add-test/another_other_dir
$ cd -
/tmp/ipfs-add-test

corntoole at cornetopia1 in /tmp/ipfs-add-test
$ tre
      1 .
      2 ├── another_dir
      3 │   ├── a.png
      4 │   └── b.png
      5 └── another_other_dir
      6     ├── c.png
      7     ├── d.png
      8     ├── foo_a.png
      9     ├── foo_b.png
     10     └── foo_c.png
     11
     12 2 directories, 7 files

corntoole at cornetopia1 in /tmp/ipfs-add-test
$ ipfs add -r -i foo*.png anot
another_dir/       another_other_dir/

corntoole at cornetopia1 in /tmp/ipfs-add-test
$ ipfs add -r -i foo*.png another_other_dir/
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH another_other_dir/c.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH another_other_dir/d.png
added QmcLiRzrFw8SMTmN9umt2EJqzij8aCGGGVqcWAhZy7j2Sj another_other_dir
  0 B / ? [---------------------------------------------------------------------------=]   0.00%
```

## Test 4 - should ignore files with the patterns `foo*.png` and `bar*.png`
```
corntoole at cornetopia1 in /tmp/ipfs-add-test
$ touch another_other_dir/bar_{e,f,g}.png

corntoole at cornetopia1 in /tmp/ipfs-add-test
$ tre
      1 .
      2 ├── another_dir
      3 │   ├── a.png
      4 │   └── b.png
      5 └── another_other_dir
      6     ├── bar_e.png
      7     ├── bar_f.png
      8     ├── bar_g.png
      9     ├── c.png
     10     ├── d.png
     11     ├── foo_a.png
     12     ├── foo_b.png
     13     └── foo_c.png
     14
     15 2 directories, 10 files

corntoole at cornetopia1 in /tmp/ipfs-add-test
$ ipfs add -r --ignore foo*.png --ignore bar*.png another_other_dir/
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH another_other_dir/c.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH another_other_dir/d.png
added QmcLiRzrFw8SMTmN9umt2EJqzij8aCGGGVqcWAhZy7j2Sj another_other_dir
 0 B / ? [---------------------------------------------------------------------------=]   0.00%

```

## Test 5 - scenario similar to icidasset's [example](https://github.com/fission-suite/go-ipfs/issues/2#issuecomment-587736794) shouldn't add any PNG files
```
corntoole at cornetopia1 in /tmp/ipfs-add-test
$ tre
      1 .
      2 ├── Drive\ Screenshots
      3 │   ├── drive-index-large.png
      4 │   └── drive-index-small.png
      5 ├── another_dir
      6 │   ├── a.png
      7 │   └── b.png
      8 ├── another_other_dir
      9 │   ├── bar_e.png
     10 │   ├── bar_f.png
     11 │   ├── bar_g.png
     12 │   ├── c.png
     13 │   ├── d.png
     14 │   ├── foo_a.png
     15 │   ├── foo_b.png
     16 │   └── foo_c.png
     17 └── small}.png
     18
     19 3 directories, 13 files

corntoole at cornetopia1 in /tmp/ipfs-add-test
$ rm small}.png

corntoole at cornetopia1 in /tmp/ipfs-add-test
$ ipfs add . -r --ignore *.png
added QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn ipfs-add-test/Drive Screenshots
added QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn ipfs-add-test/another_dir
added QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn ipfs-add-test/another_other_dir
added Qmd3eUrwsEtxC3Xvrer7Erxfj1VZeKvLGepkfKrr9PbtnU ipfs-add-test

corntoole at cornetopia1 in /tmp/ipfs-add-test
$ ~/Downloads/ipfs add . -r --ignore *.png
added QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn ipfs-add-test/Drive Screenshots
added QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn ipfs-add-test/another_dir
added QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn ipfs-add-test/another_other_dir
added Qmd3eUrwsEtxC3Xvrer7Erxfj1VZeKvLGepkfKrr9PbtnU ipfs-add-test
 0 B / ? [---------------------------------------------------------------------------=]   0.00%
corntoole at cornetopia1 in /tmp/ipfs-add-test
$ tre
      1 .
      2 ├── Drive\ Screenshots
      3 │   ├── drive-index-large.png
      4 │   └── drive-index-small.png
      5 ├── another_dir
      6 │   ├── a.png
      7 │   └── b.png
      8 └── another_other_dir
      9     ├── bar_e.png
     10     ├── bar_f.png
     11     ├── bar_g.png
     12     ├── c.png
     13     ├── d.png
     14     ├── foo_a.png
     15     ├── foo_b.png
     16     └── foo_c.png
     17
     18 3 directories, 12 files
```

## Test 6 - no ignore rules, should add all PNG files
At this point I also started testing with the binary that I  re-downloaded to make sure I didn't upload the wrong binary.
```
corntoole at cornetopia1 in /tmp/ipfs-add-test
$ ~/Downloads/ipfs add . -r
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/Drive Screenshots/drive-index-large.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/Drive Screenshots/drive-index-small.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/another_dir/a.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/another_dir/b.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/another_other_dir/bar_e.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/another_other_dir/bar_f.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/another_other_dir/bar_g.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/another_other_dir/c.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/another_other_dir/d.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/another_other_dir/foo_a.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/another_other_dir/foo_b.png
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH ipfs-add-test/another_other_dir/foo_c.png
added QmWWo6hUF5tDLUdm2bfVg7oMwW48aPx7NajCu7BgiQB4uY ipfs-add-test/Drive Screenshots
added QmQaaNSYM2RP26aoSVhgyJWEc7kMpSqaD2uKRqsZjLsK95 ipfs-add-test/another_dir
added QmUwJsnRjVABmQhtbvdZhboAFwoxknAcUvFwSPVmV1PwFD ipfs-add-test/another_other_dir
added QmSjbVRENfgCb6RXZBfxycqopa3sBPRhbgwpARX1t6NNn8 ipfs-add-test
```

**Comment by icidasset** [comment]
@corntoole That was with the binary that I downloaded from the link in the issue above (firefox send). Same thing with the `v0.5.0-dev` binary.

Thanks for the extensive tests! 🙏 

I found the issue, when I put quotes around the glob it works, but without it doesn't:
```shell
# Works
ipfs-ignore add -r --ignore "*.png" .

# Doesn't work
ipfs-ignore add -r --ignore *.png .
```

---

Detailed results:

```shell
# WITH QUOTES
~/Downloads/Drive Screenshots
⚡ ipfs-ignore add -r --ignore "*.png" .
added QmdKQbdVgBRsse2hoeq5mL9maSQJp8vzAGi9XyZbyMktmG Drive Screenshots/drive-preview-large-light.jpg
added Qmb3kRU7RAhE8Ei1FBNr6rUU9oVKqgEtHtDz53i9rgSmmn Drive Screenshots/drive-preview-small-light.jpg
added QmQWt5y3SYYAJNYGwuySTmF1gzPyqHjHkhGzB9QUb221JY Drive Screenshots

# WITHOUT QUOTES
~/Downloads/Drive Screenshots
⚡ ipfs-ignore add -r --ignore *.png .
added QmUaRffjF2jZpYJ4hZBbxXuwYimDuQPhBeJxjSoDzMZNFT Drive Screenshots/drive-index-large-dark.png
added QmTJJCkWi52uerxX9UFemX5KKPhE1aTPzwt8dsUBERzB18 Drive Screenshots/drive-index-large-light.png
added QmUXfouoNYyqWVfbwmTibSaEzbMmNRdBP27UVYJ4TayBDS Drive Screenshots/drive-index-small-dark.png
added Qme5fRH1zR133FN91SkkbGKboLnSE4WnWKEPxmxUdtzjqH Drive Screenshots/drive-index-small-light.png
added QmV9zt1UEtZUqFymDfgW417iEms1KzwpD7gH843xrwane6 Drive Screenshots/drive-preview-large-dark.png
added QmdKQbdVgBRsse2hoeq5mL9maSQJp8vzAGi9XyZbyMktmG Drive Screenshots/drive-preview-large-light.jpg
added QmTKSb1TuCBRtH5DmYGG5tdoX2xRfuYu7NbkQ4Gf1rexNq Drive Screenshots/drive-preview-small-dark.png
added Qmb3kRU7RAhE8Ei1FBNr6rUU9oVKqgEtHtDz53i9rgSmmn Drive Screenshots/drive-preview-small-light.jpg
added QmSBXWr6UQUw5ZmHaLozzRphAy7TjdYNgBgq7fynMUGeNM Drive Screenshots
added QmUaRffjF2jZpYJ4hZBbxXuwYimDuQPhBeJxjSoDzMZNFT drive-index-large-dark.png
added QmTJJCkWi52uerxX9UFemX5KKPhE1aTPzwt8dsUBERzB18 drive-index-large-light.png
added QmUXfouoNYyqWVfbwmTibSaEzbMmNRdBP27UVYJ4TayBDS drive-index-small-dark.png
added Qme5fRH1zR133FN91SkkbGKboLnSE4WnWKEPxmxUdtzjqH drive-index-small-light.png
added QmV9zt1UEtZUqFymDfgW417iEms1KzwpD7gH843xrwane6 drive-preview-large-dark.png
added QmTKSb1TuCBRtH5DmYGG5tdoX2xRfuYu7NbkQ4Gf1rexNq drive-preview-small-dark.png
```

**Comment by icidasset** [comment]
Also @corntoole, having used this, I don't seem to be able anymore to launch my original `ipfs daemon` command ...

```shell
⚡ ipfs daemon
Initializing daemon...
go-ipfs version: 0.4.23-6ce9a35
Repo version: 7
System version: amd64/darwin
Golang version: go1.13.7
20:32:09.276 ERROR   cmd/ipfs: error from node construction:  could not build arguments for function "reflect".makeFuncStub (/usr/local/Cellar/go/1.13.7/libexec/src/reflect/asm_amd64.s:12): failed to build provider.Provider: could not build arguments for function "github.com/ipfs/go-ipfs/core/node".ProviderCtor (/private/tmp/ipfs-20200130-51167-udw692/src/github.com/ipfs/go-ipfs/core/node/provider.go:24): failed to build *provider.Queue: function "github.com/ipfs/go-ipfs/core/node".ProviderQueue (/private/tmp/ipfs-20200130-51167-udw692/src/github.com/ipfs/go-ipfs/core/node/provider.go:19) returned a non-nil error: strconv.ParseUint: parsing "1582054223140642000/b": value out of range daemon.go:337

Error: could not build arguments for function "reflect".makeFuncStub (/usr/local/Cellar/go/1.13.7/libexec/src/reflect/asm_amd64.s:12): failed to build provider.Provider: could not build arguments for function "github.com/ipfs/go-ipfs/core/node".ProviderCtor (/private/tmp/ipfs-20200130-51167-udw692/src/github.com/ipfs/go-ipfs/core/node/provider.go:24): failed to build *provider.Queue: function "github.com/ipfs/go-ipfs/core/node".ProviderQueue (/private/tmp/ipfs-20200130-51167-udw692/src/github.com/ipfs/go-ipfs/core/node/provider.go:19) returned a non-nil error: strconv.ParseUint: parsing "1582054223140642000/b": value out of range
Exception: ipfs exited with 1
[tty], line 1: ipfs daemon
```

Using the `daemon` command with your binary works fine 🤔 
Any ideas?

**Comment by bmann** [comment]
@icidasset that may be because the format of the config / file storage / db changed and the new binary upgraded it automatically.

When testing, we probably want to use a custom data directory. You may need to nuke/move your current one.

`export IPFS_PATH=/path/to/ipfsrepo`

**Comment by dholms** [comment]
Everything works as expected for me (built from source on Ubuntu 18.04) :+1: 

I tested both `.gitignore` & `--ignore` functionality.

The only issue I saw is that the text for `--ignore` says:
```-i, --ignore               array  - List of ignore rules (comma separated).```

From what I can tell, this does not actually work as a comma separated list, but as a series of tags.
(ie `ipfs add -r -i test.txt -i test2.txt ./` instead of `ipfs add -r -i test.txt,test2.txt ./`

I think that functionality is fine, but the help text should reflect that.

I had similar results to @icidasset, in that fileglobs only worked when in quotes. I would say this is fine though. What do yall think?
